receiver/awsebsnvmereceiver/scraper.go

```diff
@@ -0,0 +1,200 @@
+			}
+
+			foundWorkingDevice = true
+
+			rb := s.mb.NewResourceBuilder()
+			rb.SetVolumeID(ebsDevices.volumeID)
+
+			s.recordMetric(s.mb.RecordDiskioEbsTotalReadOpsDataPoint, now, metrics.ReadOps)
+			s.recordMetric(s.mb.RecordDiskioEbsTotalWriteOpsDataPoint, now, metrics.WriteOps)
+			s.recordMetric(s.mb.RecordDiskioEbsTotalReadBytesDataPoint, now, metrics.ReadBytes)
+			s.recordMetric(s.mb.RecordDiskioEbsTotalWriteBytesDataPoint, now, metrics.WriteBytes)
+			s.recordMetric(s.mb.RecordDiskioEbsTotalReadTimeDataPoint, now, metrics.TotalReadTime)
+			s.recordMetric(s.mb.RecordDiskioEbsTotalWriteTimeDataPoint, now, metrics.TotalWriteTime)
+			s.recordMetric(s.mb.RecordDiskioEbsVolumePerformanceExceededIopsDataPoint, now, metrics.EBSIOPSExceeded)
+			s.recordMetric(s.mb.RecordDiskioEbsVolumePerformanceExceededTpDataPoint, now, metrics.EBSThroughputExceeded)
+			s.recordMetric(s.mb.RecordDiskioEbsEc2InstancePerformanceExceededIopsDataPoint, now, metrics.EC2IOPSExceeded)
+			s.recordMetric(s.mb.RecordDiskioEbsEc2InstancePerformanceExceededTpDataPoint, now, metrics.EC2ThroughputExceeded)
+			s.recordMetric(s.mb.RecordDiskioEbsVolumeQueueLengthDataPoint, now, metrics.QueueLength)
+
+			s.mb.EmitForResource(metadata.WithResource(rb.Emit()))
+		}
+
+		if foundWorkingDevice {
+			s.logger.Debug("emitted metrics for nvme device with controller id", zap.Int("controllerID", id), zap.String("volumeID", ebsDevices.volumeID))
+		} else {
+			s.logger.Debug("unable to get metrics for nvme device with controller id", zap.Int("controllerID", id), zap.String("volumeID", ebsDevices.volumeID))
+		}
+	}
+
+	return s.mb.Emit(), nil
+}
+
+// nvme0, nvme1, ... nvme{n} can have multiple devices with the same controller ID.
+// For example nvme0n1, nvme0n1p1 are all under the controller ID 0. The metrics
+// are the same based on the controller ID. We also do not want to duplicate metrics
+// so we group the devices by the controller ID.
+func (s *nvmeScraper) getEbsDevicesByController() (map[int]*ebsDevices, error) {
+	allNvmeDevices, err := s.nvme.GetAllDevices()
+	if err != nil {
+		return nil, err
+	}
+
+	devices := make(map[int]*ebsDevices)
+
+	for _, device := range allNvmeDevices {
+		deviceName := device.DeviceName()
+
+		// Check if all devices should be collected. Otherwise check if defined by user
+		hasAsterisk := s.allowedDevices.Contains("*")
+		if !hasAsterisk {
+			if isAllowed := s.allowedDevices.Contains(deviceName); !isAllowed {
+				s.logger.Debug("skipping un-allowed device", zap.String("device", deviceName))
+				continue
+			}
+		}
+
+		// NVMe device with the same controller ID was already seen. We do not need to repeat the work of
+		// retrieving the volume ID and validating if it's an EBS device
+		if entry, seenController := devices[device.Controller()]; seenController {
+			entry.deviceNames = append(entry.deviceNames, deviceName)
+			s.logger.Debug("skipping unnecessary device validation steps", zap.String("device", deviceName))
```

Comment:
nit: Some of these logs seem excessive especially if this is going to get logged multiple times during each scrape.